### PR TITLE
Restful Typeahead and initial dynamic search framework--Defunct.  For review purposes only

### PR DESCRIPTION
**Pull request by david-days**
Pull Request for code review.

### CODE REVIEW

**Comment by jywarren** [comment]
Yow, this is huge! Still catching up -- is this for code review by me?

**Comment by david-days** [comment]
We created it to use for code review, first for Ujitha and myself, and also to walk anyone (you specifically) through the changes, once we believe it's ready to present.

The PR diff page here is the tool that I was going to use to walk through on our Live discussion, and point out the differences between the one-time changes (like config/routes.rb, application.html.erb, etc) and the ongoing changes for the future (such as api.rb and typeahead.rb, etc).

Don't do anything with it for now--there are a lot of changes, and I want to make sure that they are acceptable and understandable by Ujitha and yourself before we push it in, formally.

**Comment by jywarren** [comment]
Super!

File: app/api/api.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'grape'
```
Comment (ujithaperera) [inline app/api/api.rb]: do we really need to require ?
In rails 4 we don't want. but in rails 3 I don't have a proper idea.

File: app/api/entities/doc_result.rb [hunk 2]
```diff
@@ -2,12 +2,12 @@ module API
   module Entities
     # A DocResult is an individual return item for a document (web page) search
     class DocResult < Grape::Entity
-      expose :docType
-      expose :docUrl
-      expose :docTitle
-      expose :docSummary
-      expose :docTags
-      expose :docScore
+      expose :docType, documentation: { type: "String", desc: "Type of web document being returned." }
```
Comment (ujithaperera) [inline app/api/entities/doc_result.rb]: Could you please give a brief description about this documentation parameter. I was unable to catch the real idea from the grape-entities documentation.

File: app/api/search_api.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module Search
+  class API < Grape::API
+    version 'v1', using: :header, vendor: 'publiclab'
+    format :json
+    prefix :api
+
+  resource :srch do
```
Comment (ujithaperera) [inline app/api/search_api.rb]: My resource for search is 'searches', not 'srch'. Actually I don't know whether this is a dummy resource or not. Correct me if I got it wrong.

File: lib/api/typeahead.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module API
```
Comment (ujithaperera) [inline lib/api/typeahead.rb]: why did you add this file to lib folder ?

File: lib/api/typeahead.rb [hunk 5]
```diff
@@ -1,14 +1,25 @@
 module API
   class Typeahead < Grape::API
     
-    desc 'RESTful method for typeahead for tag searches' do
-      detail 'This method conducts searches for tags that match the given value.'
-      params API::Entities::SearchRequest
-      success API::Entities::SearchResult
+    before_filter :set_search_service
+
+    resource :typeahead do
```
Comment (ujithaperera) [inline lib/api/typeahead.rb]: Don't we need this typeahead resource define in the route file before use it with grape or is it enough to have grape mounting in route file to use this resource ?

File: app/assets/javascripts/search.js [hunk 6]
```diff
@@ -1,31 +0,0 @@
...
-      })
-    },
-    updater: function(item) {
-      var url;
-      if ($(item)[0] != undefined) url = $(item)[0].attributes['data-url'].value;
-      else url = '/search/'+$('#searchform_input').val();
-      window.location = url;
-    }
-  })
-    $('#search_min_date,#search_max_date').each(function(){
```
Comment (ujithaperera) [inline app/assets/javascripts/search.js]: could you please add this small JS code to any js file in your current setup. I was unable to find this code as an addition in these changes. If I missed those lines please ignore this comment.

**Comment by ujithaperera** [comment]
Hi @david-days ,
 I have checked your changes and add few comments from my poor grape knowledge. If I asked any unnecessary question, please ignore those comments. 

And I have seen that you have removed several modifications that I have done. such as normal_search action. So I'm glad to hear our next plan from you to continue work in my end and support you with this current setup.

cheers,
Ujitha.

File: app/api/api.rb (hunk 1 above)
Comment (david-days) [inline app/api/api.rb]: I'm actually not sure.  I went through about 15-20 configuration changes to try to get this to work, and what I've posted in the PR is what functions properly and as expected.

It's possible that this (and other) "settings" are superfluous, but at this point I would recommend going with what we have and the going back to test out, line by line, any cleanup.

File: app/api/entities/doc_result.rb (hunk 2 above)
Comment (david-days) [inline app/api/entities/doc_result.rb]: The documentation parameters are used by grape-swagger to create an output that describes the RESTful service.  When running the app, you can go to "/api/swagger_doc.json" and it will put out a data structure with all the information from the documentation entries, as well as a list of all the URLs and their parameters for the endpoints.

I included grape-swagger for a couple of reasons.  First, just finding the endpoints is always a pain.  When you create any kind of web service, the application environment usually ends up dynamically setting the actual endpoints from several parameters and configuration settings.  If something is not configured properly (or it has a default value because you skipped it), then your web service may actually run just fine, but you can waste several hours just trying to guess the correct URL.  Ruby and ruby-grape is no different, and this was a way to have the system "automagically" indicate where the endpoints are.

The second reason I included grape-swagger is that it's an automated way to keep up a central "document" on what web services are available.  Once you get used to reading the documentation parameter, it becomes just like commented code.  These comments go straight into the swagger_doc output, so you won't have a mismatch between what the API is actually doing and what the documentation says.

File: app/api/search_api.rb (hunk 3 above)
Comment (david-days) [inline app/api/search_api.rb]: As I mentioned in our discussion, the "Srch" vs. "Search" usage is for namespace management.  The keyword "search" is already heavily used in the plots2 code, so we already have URL pattern matching that relies heavily on it.

Line 7 ("resource :srch do") is the line that sets the url endpoint to "/api/srch/*" for that particular set of web services.

File: lib/api/typeahead.rb (hunk 4 above)
Comment (david-days) [inline lib/api/typeahead.rb]: Hmm...I actually need to remove that.  It was one of the iterations where I was trying to get things running and configured so that it would be easy to maintain.

File: lib/api/typeahead.rb (hunk 5 above)
Comment (david-days) [inline lib/api/typeahead.rb]: This will be removed in the next PR.

File: app/assets/javascripts/search.js (hunk 6 above)
Comment (david-days) [inline app/assets/javascripts/search.js]: This set of functionality was moved to "search_form.js" and cleaned up with regards to the typeahead operations.

As I mentioned in an earlier comment, naming conventions with regards to search needed adjustment.  Since this bit of javascript actually only applies to the header search form, I renamed it. 

I will go back and make sure the "#search_min_date,#search_max_date" is there.

**Comment by david-days** [comment]
I went back through and replied to the comments so as to document our discussion from this morning.  That way, @jywarren can look through as necessary and help alleviate any confusion.

Also, you have some good catches there, especially with regards to some of the leftover garbage code that I need to clean up.  If you or anyone see anything else that you have a question about, don't be afraid to point it out--I will either answer to the best of my ability or clean up my mistake.

**Comment by ujithaperera** [comment]
@david-days ,
Thank you very much for your replies. I gathered significant knowledge from this discussion and mostly from our voice conversation. I really appreciate your hard work.

File: app/assets/javascripts/application.js [hunk 7]
```diff
@@ -29,10 +29,10 @@
 //= require leaflet/dist/leaflet
 //= require moment/moment.js
 //  APPLICATION SCRIPTS:
-//= require search.js
```
Comment (david-days) [inline app/assets/javascripts/application.js]: renamed search.js to search_form.js, as it applies to the header only

File: app/assets/javascripts/search.js (hunk 6 above)
Comment (david-days) [inline app/assets/javascripts/search.js]: datepicker formatting is now part of the _form.html.erb file, as it only applies to that form specifically.  I was having trouble gettiing advanced_search.js to work across all the versions of the search pages, so I just decided the simplest way was to include it into the form.

File: app/assets/stylesheets/application.css [hunk 8]
```diff
@@ -19,4 +19,5 @@
  *= require feature
  *= require editor
  *= require blog
+ *= require search
```
Comment (david-days) [inline app/assets/stylesheets/application.css]: search- and typeahead-specific formatting

File: app/controllers/searches_controller.rb [hunk 9]
```diff
@@ -26,6 +26,11 @@ def create
     end
   end
 
+  # Dynamic Search Page using RESTful API
```
Comment (david-days) [inline app/controllers/searches_controller.rb]: Initial "dumb" dynamic search page.  We can rework this to accept parameters and initial queries passed from other pages, once we nail down the parameter requirements.

File: app/controllers/searches_controller.rb [hunk 10]
```diff
@@ -52,10 +57,11 @@ def normal_search
                  .includes(:drupal_node_revision)
   end
 
+  # DEPRECATED 
   # utility response to fill out search autocomplete
   # needs *dramatic* optimization
-
   def typeahead
+    warn "[DEPRECATED] SearchesController.typeahead is deprecated.  Use the RESTful API for typeahead instead."
```
Comment (david-days) [inline app/controllers/searches_controller.rb]: Added a deprecation warning to old typeahead calls.  these should show up in the logs to give the admins an idea of where code needs updated.

File: app/services/search_service.rb [hunk 11]
```diff
@@ -55,8 +61,10 @@ def find_maps(input, limit=5)
         .order('nid DESC')
         .where('type = "map" AND node.status = 1 AND title LIKE ?', '%' + input + '%')
   end
-
+  
+  # DEPRECATED
   def type_ahead(id)
+    warn "[DEPRECATED] SearchService.type_ahead is deprecated.  Use the TypeaheadService methods instead."
```
Comment (david-days) [inline app/services/search_service.rb]: Added a deprecation warning to old typeahead calls. these should show up in the logs to give the admins an idea of where code needs updated.

File: app/views/layouts/_header.html.erb [hunk 12]
```diff
@@ -70,7 +70,7 @@
 
       <form id="searchform" class="navbar-form navbar-left" role="search" autocomplete="off">
         <div class="form-group">
-          <input tabindex="1" id="searchform_input" type="text" class="form-control search-query typeahead" data-provide="typeahead" placeholder="<%= t('layout._header.search') %>">
+          <input tabindex="1" id="searchform_input" type="text" class="form-control search-query typeahead" placeholder="<%= t('layout._header.search') %>"/>
```
Comment (david-days) [inline app/views/layouts/_header.html.erb]: remove-replace old typeahead calls with new functionality, implemented via REST and restful_typeahead.js

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -31,12 +31,14 @@
...
     <%= javascript_include_tag "application" %>
     <%= csrf_meta_tags %>
-
     <meta name="google-translate-customization" content="4ce4c7c384354172-5179499fc244f592-g2b333d0d29f59663-d" />
     <% if @node && @node.main_image %>
       <!-- facebook like button image: -->
       <meta property="og:image" content="<%= @node.main_image.path(:default) %>" />
     <% end %>
+
+    <%= yield(:head) %>
```
Comment (david-days) [inline app/views/layouts/application.html.erb]: One-time change to allow page-specific javascript to be loaded via the Rails "content_for (:head)" tag

File: app/views/questions/index.html.erb [hunk 14]
```diff
@@ -15,7 +15,7 @@
   <form id="questions_searchform" class="form-horizontal" role="search" autocomplete="off">
     <div class="form-group">
       <div class="col-sm-10">
-        <input tabindex="1" id="questions_searchform_input" type="text" class="form-control search-query typeahead" data-provide="typeahead" placeholder="Search or Ask a question">
+        <input tabindex="1" id="questions_searchform_input" type="text" class="form-control search-query typeahead" qryType="questions" placeholder="Search or Ask a question">
```
Comment (david-days) [inline app/views/questions/index.html.erb]: remove-replace old typeahead calls with new functionality, implemented via REST and restful_typeahead.js

File: app/views/searches/_form.html.erb [hunk 15]
```diff
@@ -1,8 +1,8 @@
 <%= form_for @search, class: 'navbar-form' do |f|%>
 
   <div class="form-group form-inline">
-    <%= f.text_field :key_words, value: params[:keyword], class: 'search-query form-control', :data => {provide: "typeahead", source: '["Balloon mapping","Thermal photography","Spectrometry"]'} %>
```
Comment (david-days) [inline app/views/searches/_form.html.erb]: remove-replace old typeahead calls with new functionality, implemented via REST and restful_typeahead.js

File: app/views/searches/_form.html.erb [hunk 16]
```diff
@@ -29,4 +29,15 @@
 
   </div>
 
-<% end %>
\ No newline at end of file
+<% end %>
+
+<script type="text/javascript">
```
Comment (david-days) [inline app/views/searches/_form.html.erb]: Added this inline javascript to handle the datepicker setup.  Since it only applies to this form, it was easier to put it here than to try to push out a separate javascript.  Not sure why, but this setup works a lot more consistently.

File: app/views/searches/advanced.html.erb [hunk 17]
```diff
@@ -1,3 +1,5 @@
+<% content_for (:head) { javascript_include_tag "advanced_search" } %>
```
Comment (david-days) [inline app/views/searches/advanced.html.erb]: This is the content_for tag to include page-specific functionality.

File: app/views/searches/advanced.html.erb [hunk 18]
```diff
@@ -7,7 +9,7 @@
   <form class="navbar-form" action="/search/advanced">
 
     <div class="form-group form-inline">
-      <input name="id" type="text" class="search-query form-control" data-provide="typeahead" data-source='["Balloon mapping","Thermal photography","Spectrometry"]' value="<%= params[:id] %>">
```
Comment (david-days) [inline app/views/searches/advanced.html.erb]: remove-replace old typeahead calls with new functionality, implemented via REST and restful_typeahead.js

File: config/application.rb [hunk 19]
```diff
@@ -1,6 +1,7 @@
 require File.expand_path('../boot', __FILE__)
 
 require 'rails/all'
+require 'rack/cors'
```
Comment (david-days) [inline config/application.rb]: one-time change to get RESTful services running.  It might be redundant with the equivalent entry from the Gemfile, but it's working for now.

File: config/application.rb [hunk 20]
```diff
@@ -63,6 +64,11 @@ class Application < Rails::Application
 
     config.action_view.sanitized_allowed_tags = ['iframe', 'embed']
 
+    # Search API
```
Comment (david-days) [inline config/application.rb]: one-time change to include the app/api folders into the applicaiton.

File: config/routes.rb [hunk 21]
```diff
@@ -1,6 +1,13 @@
 Plots2::Application.routes.draw do
 
   mount JasmineRails::Engine => '/specs' if defined?(JasmineRails)
+  
+  #Search RESTful endpoints
+  #constraints(subdomain: 'api') do
+  mount Srch::API => '/api'
```
Comment (david-days) [inline config/routes.rb]: one-time change to mount the api 'root'.  api.rb takes care of the rest of the services and endpoints.

File: config/routes.rb [hunk 22]
```diff
@@ -119,14 +126,18 @@
   match 'likes/node/:id/create' => 'like#create', :as => :add_like
   match 'likes/node/:id/delete' => 'like#delete', :as => :drop_like
 
+  #Search Pages
   match 'search' => 'searches#new'
   match 'search/advanced' => 'searches#new'
   match 'search/advanced/:id' => 'searches#new'
-  match 'search/:id' => 'searches#normal_search'
```
Comment (david-days) [inline config/routes.rb]: shuffled around the routing matches to allow for the new pages

File: config/routes.rb [hunk 23]
```diff
@@ -267,7 +278,8 @@
   # This is a legacy wild controller route that's not recommended for RESTful applications.
   # Note: This route will make all actions in every controller accessible via GET requests.
 
-  match ':controller(/:action(/:id))(.:format)'
```
Comment (david-days) [inline config/routes.rb]: According to the default comment above, this is a catch-all route match that could interfere with the RESTful API.  It's removed now, but I'm going to be writing the tests and I will keep an eye on any issues that come up.
